### DISCUSSION TITLE
Is there a way to clear refresh token cookie when ending the session ?

### DISCUSSION THREAD

**Discussion by mlbiche** [directus/directus#16161] [Archive]
Until now, our application was authenticating through the classic authentication flow with Directus. Access token, refresh token and expires were stored in the session storage. Indeed, we have the requirement to force authentication when the user ends its session (closing the tab or the browser).

We are now moving to OIDC authentication through a Keycloak server. As OIDC uses the refresh token cookie that has the same lifetime as the defined TTL value in .env, the authentication can be refreshed when the user comes back to the application on a new session. The solution to clear the refresh token cookie from the browser when the session ends would be to provide an empty expiry date in the cookie. But, it seems not possible to configure the .env this way. Affecting an empty value to `REFRESH_TOKEN_TTL` prevent the Directus server from starting.

So is there a solution to prevent authentication refresh on new application session ?

Cheers 🤙

**Comment by rijkvanzanten** [comment]
> Affecting an empty value to REFRESH_TOKEN_TTL prevent the Directus server from starting.

How did this prevent the server from starting?

FWIW I haven't tried this myself, so I don't know if this is currently possible. Technically speaking, removing the `expires` part from the cookie should turn it into a session cookie, which is what you're describing. Can you try setting `REFRESH_TOKEN_TTL="false"`?

**Reply by rijkvanzanten** [reply]
We might have to consider this a feature request if that doesn't already work 🙂

**Reply by mlbiche** [reply]
What I have tested :

- Not setting any `REFRESH_TOKEN_TTL` value in the .env config : the `expires` value correspond to 7 days (default value ?)

- Setting `REFRESH_TOKEN_TTL=""` in the .env config : The server does not start :
```bash
/Users/thibaud/Documents/Programmation/monsuivisocial/monsuivisocial-api/node_modules/ms/index.js:34
  throw new Error(
  ^

Error: val is not a non-empty string or a valid number. val=""
    at module.exports (/Users/thibaud/Documents/Programmation/monsuivisocial/monsuivisocial-api/node_modules/ms/index.js:34:9)
    at Object.<anonymous> (/Users/thibaud/Documents/Programmation/monsuivisocial/monsuivisocial-api/node_modules/directus/dist/constants.js:55:30)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Module.require (node:internal/modules/cjs/loader:1061:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/Users/thibaud/Documents/Programmation/monsuivisocial/monsuivisocial-api/node_modules/directus/dist/services/authorization.js:16:21)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)

Node.js v18.11.0
```

- Just tested, setting `REFRESH_TOKEN_TTL="false"` in the .env config : same issue pops up

I really have not dug into the Directus code, so I don't know I would be able to help you, but let me know if I can help 👍

**Reply by rijkvanzanten** [reply]
I just had another thought: You could just set the TTL to a very very low number that's still a valid cookie expires value, like `1s`. That way, the behavior is the same as a cookie without a session, but we don't have to add any special handling for a `false` case 🙂

**Reply by mlbiche** [reply]
`1s` leads to failing authentication (maybe too quick), but `5s` works properly. But using such a value leads to the incapacity to refresh the access token before it expires, as the refresh token is no longer available.

I see 2 potential solutions :
* Setting the same value for the refresh token TTL and the access token TTL (i.e. `15m`). But it would lead to an active session, even after closing the browser or tab, for the TTL duration, which could be considered as a vulnerability (?)
* Implementing a special handling for a `false` TTL value

**Reply by mlbiche** [reply]
@rijkvanzanten I could propose a contribution to implement the special handling for a `false` TTL value if needed 👍

**Reply by rijkvanzanten** [reply]
> @rijkvanzanten I could propose a contribution to implement the special handling for a `false` TTL value if needed 👍

I'll recategorize this as a feature request 👍🏻
